### PR TITLE
Avoid unhandled rejections on slow failures

### LINKED ISSUE #48: Unhandled promise rejection
Hi, 

I am using your library in one of our projects and have discovered a bug, which leads to an unhandled promise rejection.

I've investigated the problem and found the problem. The issue arises when you use retry function with timeout and custom handle timeout and the problem is on this line: https://github.com/lifeomic/attempt/blob/3da9c04a7d6bb3670df7c83ff9c21f9abfe36538/src/index.ts#L187
which leads to an error being thrown https://github.com/lifeomic/attempt/blob/3da9c04a7d6bb3670df7c83ff9c21f9abfe36538/src/index.ts#L146

The problem is that the `resolve(onError(err))` is missing a catch statement. There are multiple fixes, but I think the easiest one is to change https://github.com/lifeomic/attempt/blob/3da9c04a7d6bb3670df7c83ff9c21f9abfe36538/src/index.ts#L187 from `resolve(onError(err));` to `onError(err).then(resolve).catch(resolve);`

### PR DESCRIPTION
**Pull request by mdlavin**
Fixes #48

### CODE REVIEW

**Review by smolster — APPROVED** [review]

File: src/index.ts [hunk 1]
```diff
@@ -184,7 +184,10 @@ export async function retry<T> (
           resolve(result);
         }).catch((err: any) => {
           clearTimeout(timer);
-          resolve(onError(err));
+          // Calling resolve with a Promise that rejects here will result
+          // in an unhandled rejection. Calling `reject` with errors
+          // does not result in an unhandled rejection
+          onError(err).then(resolve).catch(reject);
```
Comment (aecorredor) [inline src/index.ts:190]: is `onError` typed as `async`?

**Review by aecorredor — APPROVED** [review]
Just curious, why we would use this library over `async-retry`?

File: src/index.ts (hunk 1 above)
Comment (mdlavin) [reply src/index.ts:190]: Yes it is - https://github.com/lifeomic/attempt/pull/50/files/39f2d292cfd63e0ef814913594e47ac7f27c12b5#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R140

**Comment by mdlavin** [comment]
> Just curious, why we would use this library over `async-retry`?

Maybe something about the personal style of this API? I'm not sure. I usually pick `async-retry`

**Pull request merged**